java/ql/src/experimental/Security/CWE/CWE-073/JFinalController.qll

```diff
@@ -0,0 +1,44 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+
+/** The class `com.jfinal.config.Routes`. */
+class JFinalRoutes extends RefType {
+  JFinalRoutes() { this.hasQualifiedName("com.jfinal.config", "Routes") }
+}
+
+/** The method `add` of the class `Routes`. */
+class AddJFinalRoutes extends Method {
+  AddJFinalRoutes() {
+    this.getDeclaringType() instanceof JFinalRoutes and
+    this.getName() = "add"
+  }
+}
+
+/** The class `com.jfinal.core.Controller`. */
+class JFinalController extends RefType {
+  JFinalController() { this.hasQualifiedName("com.jfinal.core", "Controller") }
+}
+
+/** Source model of remote flow source with `JFinal`. */
+private class JFinalControllerSource extends SourceModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "com.jfinal.core;Controller;true;getAttr" + ["", "ForInt", "ForStr"] +
```

Comment:
Thanks @atorralba for the advice, which is an excellent point. I've added a new method <code> isGetAttributeFromRemoteSource</code> and relevant taint configuration & classes to handle this scenario - a value is retrieved from a remote source, which is likely to be a <code>getPara</code> call, then passed to a request/session attribute setter, then passed to a request/session attribute getter. Please let me know if there is a better way to model this.

Reply:
Using `localFlow` in a source definition can cause performance problems because of recursion. I think the proper way to implement this would be through taint steps, i.e. `setAttr*` taints the session object of the JFinal controller (which could be a synthetic field), and `getAttr*` returns that taint. However, the fact that this hasn't been done for `HttpServletRequest` and `HttpSession` makes me believe it may be difficult to do correctly (for instance, `HttpSession s = getSession(); s.setAttribute("attr", taint())` and `HttpSession s = getSession(); s.getAttribute("attr")` can't be wired in that way). Worth a shot if you want — alternatively, I'd just remove the `getAttr` and `getSession` sources.

Reply:
I've changed taint configuration for request/session attribute getters/setters to CSV data model.

However, for session object access e.g.
```
		getSession().setAttribute("uploadDir", savePath);
		String sessionUploadDir = getSessionAttr("uploadDir");
```
The following doesn't help:
```
        "com.jfinal.core;Controller;true;getSession;;;SyntheticField[com.jfinal.core.Controller.session] of Argument[-1];ReturnValue;value",
        "javax.servlet.http;HttpSession;true;setSessionAtribute;;;Argument[0];MapKey of SyntheticField[com.jfinal.core.Controller.session] of Argument[-1];value",
        "javax.servlet.http;HttpSession;true;setSessionAtribute;;;Argument[1];MapValue of SyntheticField[com.jfinal.core.Controller.session] of Argument[-1];value"
```

Maybe the problem is caused by that <code>HttpSession</code> is an interface or some internal implementation of CodeQL doesn't support it? Or what could be wrong with my implementation? I've added a test case for this scenario and marked it as not detected by the current query.

Thanks,
@luchua-bc

Reply:
Let's leave it as it is if the framework support can be available in a release in the near future. Otherwise I can change it back to plain CodeQL with a taint configuration that checks attribute names match. The former is probably better since it's more efficient and concise.

Reply:
I don't think there's plans of supporting JFinal in the short term. So this PR will probably be all there is about it (at least by now). I agree with leaving it as it is in this PR, but anyway if you decided to change it, you wouldn't need another taint configuration to check the names — probably local taint would suffice.

Reply:
I see. I will try local taint to make it check the names and be fully functional.

Reply:
Hmm, you went back to square one modeling this as a source :) As I mentioned above, I think this works better as a taint step (or better, an `AdditionalValueStep` since the value is preserved), but with plain CodeQL and not CSV summaries.

Reply:
Almost there! When using a class extending `AdditionalValueStep` you don't need to add it to the specific taint tracking configuration: it's picked up by default and applied to all configurations. So you don't need to override `isAdditionalTaintStep`.

Reply:
Thanks @atorralba. I've removed the unnecessary <code>isAdditionalTaintStep</code>.